vowpalwabbit/example_predict.h

```diff
@@ -42,7 +43,8 @@ struct example_predict
 
   v_array<namespace_index> indices;
   std::array<features, NUM_NAMESPACES> feature_space;  // Groups of feature values.
-  uint64_t ft_offset;                                  // An offset for all feature values.
+  uint64_t ft_offset;                                  // SAn offset for all feature values.
+  uint32_t stack_depth;
```

Comment:
why is stack_depth needed as part of this class, not following that clearly but intuitively feels like coupling of reduction stack with example

Reply:
I use VW_DBG to debug learn/predict and show state of reductions in the context of processing a given example.  Tracking this in example as it traverses the stack seemed like a good way to do it.

Reply:
+1 to this

Reply:
The fact that a 'debug build' debug statements now force abstractions to get coupled with other concepts of other abstractions is a hard sell for me. I suggest you either flesh out a design doc about debug statements to see how to address it or remove it and leave it for ad-hoc debugging since it's been helpful for you.

Reply:
This works in debug and release mode.

I view this as general low impact trace facility for developing reductions.  The lack of trace facility in reductions was a substantial issue both in developing continuous actions and with predict/learn changes.  

There are other changes planned to predict/learn in the near future where this will be important.  One of these changes is that learn() can avoid a substantial number of predictions in library mode.    

Perhaps best discussed during the VW session as I know this has been a point of contention in the past.